feature/working-memory-dry

thomashebrard opened the pull request:
### 📝 Description

## Feature

- `WorkingMemoryFactory`: added method `make_for_dry_run`: Takes an array of tuple (`stuff_name`, `concept_code`, `class_structure`) and generates with `polyfactory` a working memory with fake objects.
- Completed dry run for `PipeCondition`, `PipeParallel`, `PipeBatch`. Added static validation for those as well. 
- Added a config in the DryRunConfig: `allowed_to_fail_pipes` enables to list pipes that are allowed to fail the dry run, for example the pipe explicitly coded for creating an infinite loop 
- `ConceptLibrary`: Added method `is_concept_code_legal`, that checks that the domain attached to the concept exists.
- Added `pipe_type: PipeType` to pipes classes (CONTROLLER/OPERATOR)
- `PipeLibrary`:  added method `dry_run_all_pipes` that runs the dry_run for all pipes.
- Added validation of TOML files: Raise an error when there are trailing whitespaces, compilation blockers...

## Fix

- Removed useless prints

### 🔄 Type of Change
- [ ] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 💥 Breaking change
- [ ] 📚 Documentation update
- [ ] 🧹 Code refactor
- [ ] ⚡ Performance improvement
- [ ] ✅ Test update

### 🧪 Tests

### 📋 Checklist

- [ ] Linting / type-checking / unit tests pass locally with my changes
- [ ] I've added tests that prove my fix is effective or that my feature works
- [ ] My code follows the style guidelines of this project
- [ ] I've made corresponding changes to the documentation
- [ ] My changes generate no new warnings

thomashebrard committed:
feature/working-memory-dry

thomashebrard committed:
add toml validation

thomashebrard committed:
rename folder test_pipelines

thomashebrard committed:
fix tests

thomashebrard committed:
fix toml test

thomashebrard committed:
fix combined stuff PipeParallel

thomashebrard committed:
ok

thomashebrard committed:
order testsadding tests

thomashebrard committed:
fix lint

thomashebrard committed:
fix tests

thomashebrard committed:
make cc

thomashebrard committed:
Fix all needed inputs

thomashebrard committed:
make cc

thomashebrard committed:
fix poluyfactory Optionals

thomashebrard committed:
remove prints

thomashebrard committed:
remove prints

thomashebrard committed:
rename folder

thomashebrard committed:
Merge branch 'dev' into feature/working-memory-dry

thomashebrard committed:
fix preliminary_text

thomashebrard committed:
put as default the feature flags

thomashebrard committed:
Adding cursor rules

thomashebrard committed:
run pipelix valdiate

thomashebrard committed:
new cursor rules

thomashebrard committed:
remove prints

thomashebrard committed:
fix/cursor-rules

thomashebrard committed:
remove double brackets

thomashebrard committed:
make cc

thomashebrard committed:
Merge pull request #138 from Pipelex/fix/cursor-rules

Fix/cursor rules

thomashebrard committed:
Merge branch 'dev' into feature/working-memory-dry

thomashebrard committed:
make cc

thomashebrard committed:
fix PIPE_OCR_INPUT_NAME

thomashebrard committed:
make cc

thomashebrard committed:
make cc

thomashebrard committed:
fix tests

thomashebrard committed:
fix make t

thomashebrard committed:
fix make t

thomashebrard committed:
fix rules

thomashebrard committed:
Merge pull request #139 from Pipelex/staging

fix rules

thomashebrard committed:
fix PIpeBatch

thomashebrard committed:
make cc

thomashebrard committed:
make cc

thomashebrard committed:
make cc

thomashebrard committed:
make cc

lchoquel committed:
Feature/working memory dry 2 (#140)

* Moved   dry_run_all_pipes() out of PipeLibrary, it's a static func called (or not) by make()

* Added dry_run_pipes in its noew module

* Nicer log.info with indentation when running pipes

* Improve Pipe log info indent

thomashebrard committed:
pipeCondition

thomashebrard committed:
test_boot

thomashebrard committed:
add test boot

thomashebrard committed:
end of line

thomashebrard committed:
Merge pull request #141 from Pipelex/feature/test-boot

add test boot

lchoquel commented on .cursor/rules/best_practices.mdc:
```diff
@@ -75,3 +71,7 @@ This document outlines the core best practices and patterns used in our codebase
+
+## Pipelines
+
+Always run the cli `pipelex validate` when you are finished writing pipelines: This checks for errors. If there are errors, iterate.
```
`make validate` otherwise Cursor runs it outside the .venv every other time and it fails

lchoquel commented on .cursor/rules/pipe-batch.mdc:
```diff
@@ -0,0 +1,31 @@
+---
+# PipeBatch Controller
+
+The PipeBatch controller allows you to apply a pipe operation to each element in a list of inputs in parallele. It is created via a PipeSequence.
```
in parallel.

lchoquel commented on .cursor/rules/pipe-batch.mdc:
```diff
@@ -0,0 +1,31 @@
+
+# Important tip
+
+Always run the cli `pipelex validate` when you are finished writing pipelines: This checks for errors. If there are errors, iterate.
```
same

lchoquel commented on .cursor/rules/pipe-condition.mdc:
````diff
@@ -0,0 +1,50 @@
+
+```toml
+[pipe.conditional_operation]
+PipeCondition = "A conditonal pipe to decide wheter..."
````
whether

lchoquel commented on .cursor/rules/pipe-condition.mdc:
```diff
@@ -0,0 +1,50 @@
+
+- `expression`: Direct boolean or string expression (mutually exclusive with expression_template)
+- `expression_template`: Jinja2 template for more complex conditional logic (mutually exclusive with expression)
+- `pipe_map`: Dictionary mapping expression results to pipe codes : 
```
to pipe codes:

lchoquel commented on .cursor/rules/pipe-condition.mdc:
```diff
@@ -0,0 +1,50 @@
+- `expression_template`: Jinja2 template for more complex conditional logic (mutually exclusive with expression)
+- `pipe_map`: Dictionary mapping expression results to pipe codes : 
+1 - The key on the left (`small`, `medium`) is the result of `expression` or `expression_template`.
+2 - The value on the right (`process_small`, `process_medium`, ..) is the name of the pipce to trigger
```
of the pipe to trigger

lchoquel commented on .cursor/rules/pipe-condition.mdc:
```diff
@@ -0,0 +1,50 @@
+
+# Important tip
+
+Always run the cli `pipelex validate` when you are finished writing pipelines: This checks for errors. If there are errors, iterate.
```
same

lchoquel commented on .cursor/rules/pipe-llm.mdc:
```diff
@@ -0,0 +1,107 @@
+[pipe.write_story]
+PipeLLM = "Write a short story"
+output = "Text"
+prompt_template = """
```
not a template, just a prompt

lchoquel commented on .cursor/rules/pipe-llm.mdc:
````diff
@@ -0,0 +1,107 @@
+```toml
+[pipe.analyze_image]
+PipeLLM = "Analyze image"
+inputs = { image = "Image" } # `image` is the name of the stuff that contains the Image. If its in a stuff, you can add something like `{ "page.image": "Image" }
````
If it's

lchoquel commented on .cursor/rules/pipe-llm.mdc:
```diff
@@ -0,0 +1,107 @@
+
+# Important tip
+
+Always run the cli `pipelex validate` when you are finished writing pipelines: This checks for errors. If there are errors, iterate.
```
same

lchoquel commented on .cursor/rules/pipe-ocr.mdc:
```diff
@@ -0,0 +1,39 @@
+globs: 
+alwaysApply: false
+---
+# PipeOCR Guide
```
It's `PipeOcr`

lchoquel commented on .cursor/rules/pipe-ocr.mdc:
```diff
@@ -0,0 +1,39 @@
+
+# Important tip
+
+Always run the cli `pipelex validate` when you are finished writing pipelines: This checks for errors. If there are errors, iterate.
```
same

lchoquel commented on .cursor/rules/pipe-sequence.mdc:
```diff
@@ -0,0 +1,58 @@
+
+# Important tip
+
+Always run the cli `pipelex validate` when you are finished writing pipelines: This checks for errors. If there are errors, iterate.
```
same

lchoquel commented on .cursor/rules/pipes.mdc:
```diff
@@ -1,206 +1,80 @@
 
-## Pipeline file naming
+- Always first write your "plan" in natural langage, then transcribe it in pipelex.
+- Always run `pipelex validate` before finishing a task. It will ensure the pipe is runnable
```
same

lchoquel commented on .cursor/rules/pipes.mdc:
```diff
@@ -1,206 +1,80 @@
+- Never use plurals (no "Stories", use "Story")
+- Avoid adjectives (no "LargeText", use "Text")
+- Don't redefine native concepts (Text, Image, PDF, TextAndImages, Number)
+yes 
```
typo?

lchoquel commented on .cursor/rules/pipes.mdc:
```diff
@@ -1,206 +1,80 @@
-- Naming: pipe names are snake_case, like python functions (e.g., `extract_gantt_tasks`)
-- Define pipes using as PipeLLM, PipeSequence, PipeParallel, PipeCondition
+The pipes will all have at least this base structure. 
+- `inputs`: Dictionnary of key behing the variable used in the prompts, and the value behing the ConceptName
```
pipes can exist without inputs

lchoquel replied:
I don't understand "behind" here

lchoquel commented on .cursor/rules/pipes.mdc:
```diff
@@ -1,206 +1,80 @@
-- Define pipes using as PipeLLM, PipeSequence, PipeParallel, PipeCondition
+The pipes will all have at least this base structure. 
+- `inputs`: Dictionnary of key behing the variable used in the prompts, and the value behing the ConceptName
+- `output`: The name of the concept to output. The `ConceptName` should have the same name as the python class if you want structured output. Look on how to do that here: [structures.mdc](structures.mdc)
```
Look how

lchoquel commented on .cursor/rules/pipes.mdc:
```diff
@@ -1,206 +1,80 @@
+## Pipe Controllers and Pipe Operator
 
-Get inspiration from this template filled with appropriate examples:
+Look at the Pipes we have in order to adapt it. Pipes are organized in two categories:
```
adapt them

lchoquel commented on .cursor/rules/pipes.mdc:
```diff
@@ -1,206 +1,80 @@
-- Write the pipe's toml and required structure in pipelex/pipelex_libraries/pipelines
+# Important tip
 
+Always run the cli `pipelex validate` when you are finished writing pipelines: This checks for errors. If there are errors, iterate.
```
same

lchoquel commented on .cursor/rules/prompt-templates.mdc:
```diff
@@ -80,5 +71,5 @@ Be sure to make the proper choice of prefix for each insertion.
 ## Style rule
 
-Always favor the @ and $ prefixes instead of the jinja2 double-curly-brackets syntax, unless explicitly instructed otherwise.
+If you need to add more complexe jinja2 template, we support the classic double curly brackets `{{ }}`
```
complex

lchoquel commented on .cursor/rules/structures.mdc:
```diff
@@ -25,7 +25,7 @@ from pydantic import BaseModel, Field
 from pipelex.core.stuff_content import StructuredContent
 
-class YourModel(StructuredContent):
+class YourModel(StructuredContent): # Always be a subclass of StructuredContent
```
Must always be

lchoquel commented on Makefile:
```diff
@@ -285,6 +285,14 @@ test-inference: env
+ti-dry: env
+	$(call PRINT_TITLE,"Unit testing")
+	@if [ -n "$(TEST)" ]; then \
+		$(VENV_PYTEST) --pipe-run-mode dry --exitfirst -m "inference and not imgg" -s -k "$(TEST)" $(if $(filter 1,$(VERBOSE)),-v,$(if $(filter 2,$(VERBOSE)),-vv,$(if $(filter 3,$(VERBOSE)),-vvv,))); \
```
no point in skipping imgg here

lchoquel commented on docs/pages/build-reliable-ai-workflows-with-pipelex/pipe-controllers/PipeCondition.md:
```diff
@@ -22,73 +22,126 @@ The `PipeCondition` controller adds branching logic to your pipelines. It evalua
+output = "native.Text"
+expression = "input_data.category"
+
+[pipe.route_by_category.pipe_map]
```
not a good example of categorization, use something that makes more sense like fiction/news/technical

lchoquel commented on docs/pages/build-reliable-ai-workflows-with-pipelex/pipe-controllers/PipeCondition.md:
```diff
@@ -22,73 +22,126 @@ The `PipeCondition` controller adds branching logic to your pipelines. It evalua
+output = "native.Text"
 prompt_template = """
-Classify the document as 'invoice' or 'receipt'
+Output this only: "small"
```
I hate bogus examples like these

lchoquel commented on docs/pages/build-reliable-ai-workflows-with-pipelex/pipe-controllers/PipeCondition.md:
```diff
@@ -22,73 +22,126 @@ The `PipeCondition` controller adds branching logic to your pipelines. It evalua
-[pipe.process_receipt]
-# ... pipe definition ...
-output = "ProcessedDocument"
+How this works:
```
How this particular example works:

lchoquel commented on docs/pages/build-reliable-ai-workflows-with-pipelex/pipe-controllers/PipeCondition.md:
```diff
@@ -22,73 +22,126 @@ The `PipeCondition` controller adds branching logic to your pipelines. It evalua
+inputs = { classification = "DocumentType" }
+output = "ProcessedDocument"
+prompt_template = """
+Process this invoice document...
```
it would look better with actual (simplified) example from our invoice processor

lchoquel commented on docs/pages/build-reliable-ai-workflows-with-pipelex/pipe-controllers/PipeCondition.md:
```diff
@@ -22,73 +22,126 @@ The `PipeCondition` controller adds branching logic to your pipelines. It evalua
+"""
+
+[pipe.process_unknown]
+PipeLLM = "Handle unknown document types"
```
this is confusing. A document type is .pdf, .doc etc.

lchoquel commented on pipelex/core/concept_provider_abstract.py:
```diff
@@ -1,10 +1,15 @@
+
 
 class ConceptProviderAbstract(ABC):
+    root: ConceptLibraryRoot = Field(default_factory=dict)
```
this defeats the purpose of the Abstract class, it's part of implementation, should not be needed

lchoquel replied:
🛑

lchoquel commented on pipelex/core/pipe_input_spec.py:
```diff
@@ -56,6 +57,14 @@ def get_required_concept_code(self, variable_name: str) -> str:
+        return cls(root={})
+
+    @classmethod
+    def make_from_dict(cls, input_dict: Dict[str, str]) -> Self:
```
the type is named, use it `PipeInputSpecRoot`

lchoquel commented on pipelex/core/pipe_provider_abstract.py:
```diff
@@ -1,10 +1,15 @@
+
 
 class PipeProviderAbstract(ABC):
+    root: PipeLibraryRoot = Field(default_factory=dict)
```
defeats the purpose of the Abstract

lchoquel replied:
🛑

lchoquel commented on pipelex/core/working_memory_factory.py:
```diff
@@ -120,3 +123,57 @@ def make_from_memory_file(cls, memory_file_path: str) -> WorkingMemory:
+        for variable_name, concept_code, structure_class in needed_inputs:
+            log.debug(f"Creating dry run mock for '{variable_name}' with concept '{concept_code}' and class '{structure_class.__name__}'")
+
+            try:
```
declare mock_content beforehand:
`mock_content: StuffContent`

lchoquel commented on pipelex/core/working_memory_factory.py:
```diff
@@ -120,3 +123,57 @@ def make_from_memory_file(cls, memory_file_path: str) -> WorkingMemory:
+                working_memory.add_new_stuff(name=variable_name, stuff=mock_stuff)
+
+            except Exception as e:
+                log.warning(f"Failed to create mock for '{variable_name}' ({concept_code}): {e}. Using fallback text content.")
```
why did it fail? if it does happen, then catch the exact Exception and make sense of it

lchoquel commented on pipelex/exceptions.py:
```diff
@@ -169,6 +169,15 @@ class PipeStackOverflowError(PipeExecutionError):
+    """Raised when a dry run fails due to missing inputs or other validation issues."""
+
+    def __init__(self, message: str, missing_inputs: Optional[List[str]] = None, pipe_code: Optional[str] = None):
+        self.missing_inputs = missing_inputs or []
```
exlicit typing:
` self.missing_inputs: List[str] = missing_inputs or []`

lchoquel commented on pipelex/pipe_controllers/pipe_condition.py:
```diff
@@ -155,3 +291,89 @@ async def _run_controller_pipe(
+            try:
+                get_required_pipe(pipe_code=pipe_code)
+                log.debug(f"Pipe '{pipe_code}' exists and is accessible")
+            except Exception as e:
```
for all cases of `Exception as e:` replace by `Exception as exc:`

lchoquel commented on pipelex/pipe_controllers/pipe_condition.py:
```diff
@@ -155,3 +291,89 @@ async def _run_controller_pipe(
+        missing_pipes: List[str] = []
+        for pipe_code in all_pipe_codes:
+            try:
+                get_required_pipe(pipe_code=pipe_code)
```
- either keep those pipes in a dict to avoid having to `get_required_pipe ` again below
- or merge the process, fry run will require all pipes so the error will happen in the loop below anyway

lchoquel commented on pipelex/pipe_controllers/pipe_condition.py:
```diff
@@ -155,3 +291,89 @@ async def _run_controller_pipe(
+                f"Dry run failed for pipe '{self.code}' (PipeCondition): missing pipes: {', '.join(missing_pipes)}. "
+                f"Pipe map: {self.pipe_map}, default: {self.default_pipe_code}"
+            )
+            raise DryRunError(
```
this error class is too generic
it should be a `DryRunPipeNotFoundError` and `DryRunPipeNotFoundError` should be a subclass of `DryRunError`

lchoquel commented on pipelex/pipe_controllers/sub_pipe.py:
```diff
@@ -55,20 +55,42 @@ async def run(
-                output_name=self.output_name,
-            )
+            # This is the only line that changes between run and dry_run
+            if sub_pipe_run_params.run_mode == PipeRunMode.DRY:
```
always use match/case for enums like `PipeRunMode`

lchoquel commented on pipelex/pipe_controllers/sub_pipe.py:
```diff
@@ -55,20 +55,42 @@ async def run(
-                pipe_run_params=sub_pipe_run_params,
-            )
+            # This is the only line that changes between run and dry_run
+            if sub_pipe_run_params.run_mode == PipeRunMode.DRY:
```
same

lchoquel commented on pipelex/pipe_controllers/sub_pipe.py:
```diff
@@ -81,13 +103,24 @@ async def run(
-                pipe_run_params=sub_pipe_run_params,
-            )
+            # This is the only line that changes between run and dry_run
+            if sub_pipe_run_params.run_mode == PipeRunMode.DRY:
```
same

lchoquel commented on pipelex/pipe_operators/pipe_llm_prompt.py:
```diff
@@ -145,12 +147,13 @@ async def _run_operator_pipe(
-                    image_uri = prompt_image_content.url
-                    user_image = PromptImageFactory.make_prompt_image_from_uri(uri=image_uri)
-                prompt_user_images.append(user_image)
+                if prompt_image_content is not None:  # An ImageContent can be optional..
```
not it can't be, we're examining `for user_image_name in self.user_images:` so if we have a user_image_name, it must be found in the working memory above

lchoquel commented on pipelex/pipe_operators/pipe_ocr.py:
```diff
@@ -229,7 +239,10 @@ async def _dry_run_operator_pipe(
     ) -> PipeOutput:
-        log.info(f"PipeOcr: dry run operator pipe: {self.code}")
+        log.debug(f"PipeOcr: dry run operator pipe: {self.code}")
+        if pipe_run_params.run_mode != PipeRunMode.DRY:
```
awlways use match/case

lchoquel commented on pipelex/pipelex.py:
```diff
@@ -254,11 +259,14 @@ def teardown(self):
         log.info(f"Pipelex {PACKAGE_VERSION} initialized.")
+        if dry_run_all_pipes:
+            all_pipes = pipelex_instance.library_manager.pipe_library.get_pipes()
+            asyncio.run(dry_run_pipes(pipes=all_pipes))
```
add  a `# TODO: move this somewhere we can make a proper async call`

lchoquel commented on pipelex/tools/misc/toml_utils.py:
```diff
@@ -1,17 +1,122 @@
+        _validate_toml_content(content, path)
+
+
+def _clean_trailing_whitespace(content: str) -> str:
```
cleanup

lchoquel commented on pipelex/tools/misc/toml_utils.py:
```diff
@@ -1,17 +1,122 @@
+    3. If no empty line at EOF, removes trailing whitespace from last non-empty line
+
+    Args:
+        content: The TOML content to clean
```
cleanup

lchoquel commented on pipelex/tools/misc/toml_utils.py:
```diff
@@ -1,17 +1,122 @@
+
+        # If content changed, write it back
+        if content != cleaned_content:
+            with open(path, "w", encoding="utf-8") as file:
```
writing fixed file should be an option: `is_apply_fixes_enabled`
this must be passed down from the top CLI call

lchoquel replied:
also, the method name is not appropriate for this process, maybe separate check (with/without fix) and load

lchoquel commented on pipelex/tools/misc/toml_utils.py:
```diff
@@ -1,17 +1,122 @@
+    try:
+        return load_toml_from_path(path)
+    except toml.TomlDecodeError as exc:
+        print(f"Failed to parse TOML file '{path}': {exc}")
```
no, the point of `failable_load_toml_from_path` is to gracefully fail and return None **only in case of FileNotFound.**
a DecodeError should raise
and clearly we should rename `failable_load_toml_from_path` to match the idea, maybe `load_toml_from_path_if_exists`

lchoquel commented on tests/conftest.py:
```diff
@@ -33,17 +33,15 @@ def reset_pipelex_config_fixture():
-    print("\n")
     yield
     # Code to run after each test
-    print("\n")
```
all those "\n" are here for a purpose

lchoquel commented on tests/e2e/pipelex/pipes/pipe_controller/pipe_condition/test_pipe_condition_1.py:
```diff
@@ -0,0 +1,58 @@
+from tests.test_pipelines.pipe_controllers.pipe_condition.pipe_condition import CategoryInput
+
+
+@pytest.mark.llm
```
when calling pipes, never use marker `llm`, because you can't know what the pipe will do, just use `inference`

lchoquel commented on tests/e2e/pipelex/pipes/pipe_controller/pipe_condition/test_pipe_condition_1.py:
```diff
@@ -0,0 +1,58 @@
+        # Basic assertions
+        assert pipe_output is not None
+        assert pipe_output.working_memory is not None
+        if pipe_run_mode != PipeRunMode.DRY:
```
always use match/case

lchoquel commented on tests/e2e/pipelex/pipes/pipe_controller/pipe_condition/test_pipe_condition_2.py:
```diff
@@ -0,0 +1,58 @@
+from tests.test_pipelines.pipe_controllers.pipe_condition.pipe_condition import CategoryInput
+
+
+@pytest.mark.llm
```
same

lchoquel commented on tests/e2e/pipelex/pipes/pipe_controller/pipe_condition/test_pipe_condition_2.py:
```diff
@@ -0,0 +1,58 @@
+        # Basic assertions
+        assert pipe_output is not None
+        assert pipe_output.working_memory is not None
+        if pipe_run_mode != PipeRunMode.DRY:
```
same

lchoquel commented on tests/e2e/pipelex/pipes/pipe_controller/pipe_condition/test_pipe_condition_simple.py:
```diff
@@ -0,0 +1,89 @@
+            # If it succeeds completely
+            assert pipe_output is not None
+            assert pipe_output.working_memory is not None
+            print("✅ Direct PipeCondition SUCCEEDED completely!")
```
replace prints with logs or just remove them

lchoquel commented on tests/e2e/pipelex/pipes/pipe_controller/pipe_sequence/test_pipe_sequence_batching.py:
```diff
@@ -0,0 +1,74 @@
+async def test_review_analysis_sequence_with_batching(pipe_run_mode: PipeRunMode):
+    """Test customer review analysis sequence with batching."""
+    # Create test input - a document with reviews
+    if pipe_run_mode == PipeRunMode.DRY:
```
same

lchoquel commented on tests/integration/pipelex/pipes/pipe_controller/pipe_condition/test_pipe_condition_complex.py:
```diff
@@ -0,0 +1,293 @@
+
+@pytest.mark.dry_runnable
+@pytest.mark.inference
+@pytest.mark.llm
```
remove mark llm

lchoquel commented on tests/integration/pipelex/pipes/pipe_controller/pipe_condition/test_pipe_condition_complex.py:
```diff
@@ -0,0 +1,293 @@
+        pretty_print(pipe_output, title="Business Finance Processing")
+
+        assert pipe_output is not None
+        if pipe_run_mode != PipeRunMode.DRY:
```
same

lchoquel reviewed (changes requested):
Only 2 blocking remarks, marked 🛑

lchoquel committed:
Fix/abstract providers vs concrete libraires (#142)

* fix/Abstract providers vs Concrete libraires WIP

* Fix Abstract libraries thanks to LibraryManagerAbstract

* Separate Pipelex make() from validate_libraries() and dry_run_pipes()
TestFundamentals runs the tests separately, called by gha tests_check.yml before the other tests.

* Simplify calls to dry_run_all_pipes()

* Make ConceptProviderAbstract only a provider, like before

* Make DomainProviderAbstract and PipeProviderAbstract only providers, like before

* Rename failing_pipelines

thomashebrard committed:
make cc

thomashebrard committed:
remove thread tests in gha

lchoquel committed:
Disable pytest-xdist for `make gha-tests` and `make tp TEST=TestFundamentals`
